Create ADR for installing only selected CRDs

matthchr opened the pull request:
Related to #2920.

**If applicable**:
- [x] this PR contains documentation
- [ ] this PR contains tests

theunrepentantgeek commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,174 @@
+As mentioned above, a key goal is a reduction in the default number of CRDs installed. That means we somehow need to 
+choose which CRDs we will install and which CRDs we will not.
+
+#### Answer 1a: We handpick a set of resources we feel are "critical" and install just those by default
+
```
```suggestion
#### Answer 1a: Handpicked Default

We handpick a set of resources we feel are "critical" and install just those by default

```

matthchr replied:
Done

theunrepentantgeek commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,174 @@
+* May result in higher kube-apiserver memory consumption than is strictly needed for users as it is likely to include some
+  resources which they may not use.
+
+#### Answer 1b: We install nothing by default
+
```
```suggestion
#### Answer 1b: Nothing by default

We install nothing by default

```

theunrepentantgeek commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,174 @@
+      Or maybe it needs to actually upgrade those versions too even though they weren't specified, to maintain a consistent
+      storage version?
+
+### Question 1: How to identify the subset of resources to install by default?
```
```suggestion
### Question 1: What to install by default?
```

theunrepentantgeek commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,174 @@
+Cons:
+* The operator is nonfunctional out of the box, which is a poor new-user experience.
+
+#### Answer 1c: Selection of CRDs to install is required and the operator pod will not launch without it
```
```suggestion
#### Answer 1c: Mandatory configuration

The selection of CRDs to install is required and the operator pod will not launch without it
```

theunrepentantgeek commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,174 @@
+* Exists today
+
+Cons:
+* Command-line isn't preserved between upgrades, which fails requirement #3 (seamless upgrade experience).
```
```suggestion
* Command-line isn't preserved between upgrades, which may fail requirement #3 (seamless upgrade experience).
```

theunrepentantgeek commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,174 @@
+[pre-install hook](https://stackoverflow.com/questions/55503893/helm-patch-default-service-account). We would create and
+patch the `ConfigMap` in the pre-install hook.
+
+ConfigMap structure:
```
```suggestion
One possible configMap structure:
```

theunrepentantgeek commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,174 @@
+If nothing is specified as part of the upgrade, the set of supported resources will not change.
+
+If the `ConfigMap` doesn't exist when we go to upgrade, we can assume we're coming from an older ASO version that didn't
+support the `ConfigMap` and default to creating one that signals to install all resources.
```
This would put us back into the high memory use scenario.

matthchr replied:
Yes but it'd be a breaking change IMO to do anything other than this, right?

theunrepentantgeek replied:
Yup. I just figure it's worth calling out as a Con for this approach.

matthchr replied:
Note I'm specifically here referring to just upgrade scenarios. New installs user would need to choose still. I'll call that out as well.

matthchr replied:
Added the con

theunrepentantgeek commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,174 @@
+
+### Question 2: How do users specify what CRDs to install?
+
+#### Answer 2a: using our existing --crd-pattern command line parameter on the operator pod
```
```suggestion
#### Answer 2a: Use the existing command line parameter

Continue using the existing `--crd-pattern` command line parameter on the operator pod.
```

theunrepentantgeek commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,174 @@
+Cons:
+* Command-line isn't preserved between upgrades, which fails requirement #3 (seamless upgrade experience).
+
+#### Answer 2b: Using a ConfigMap
```
```suggestion
#### Answer 2b: Use a ConfigMap
```

theunrepentantgeek commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,174 @@
+We investigated doing this originally for [Helm chart size limitations](../ADR-2023-02-Helm-Chart-Size-Limitations)
+
+Pros:
+* Slightly more structured than a `ConfigMap`.
```
```suggestion
* Significantly more structured than a `ConfigMap`.
```

matthchr replied:
Dropped the qualifier and said "More structured than a `ConfigMap`"

theunrepentantgeek commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,174 @@
+## Decision
+
+For question #1: [Answer 1c: Selection of CRDs to install is required and the operator pod will not launch without it](#answer-1c--selection-of-crds-to-install-is-required-and-the-operator-pod-will-not-launch-without-it)
+For question #2: [Answer 2b: Using a ConfigMap](#answer-2b--using-a-configmap)
```
With the separation of question 2 into "how to configure" and "how to ensure consistency" (as mentioned elsewhere), I think we take a different approach:

* Mandatory configuration
* Retain Command line parameter
* Always support CRDs that have already been installed

matthchr replied:
Not sure exactly what "Retain command line parameter" means - does it just mean "use the option we have now" or are you talking about some cmdline parameter that is retained between deployments somehow automatically?

I agree with your argument towards "Always support CRDs that have already been installed", but I think that actually forces us away from "Mandatory configuration", at least in the sense of Helm required parameters or anything like that.

I think what we want is:
- Operator exits with a clear reason (visible on Pod status) if no CRDs are configured.
- Cmdline parameter to configure which _new_ CRDs to install
- All existing CRDs are automatically upgraded to the latest version, even if not specified in cmdline parameter.

If we have some form of mandatory configuration _and_ the "existing CRDs" behavior it's not clear what the user should put into the mandatory configuration if they want to just use the resources they were already using.

theunrepentantgeek reviewed (changes requested):
A good write up of the problem, but I disagree with your conclusion.

I think your answers for Question 2 are actually trying to answer two different questions:
* How do users configure which CRDs to install
* How do we ensure consistent configuration as ASO is upgraded?

For the later question, there are at least four answers:

1. Require the user to be consistent when installing any new version 
   * Pro: Easy for us
   * Con: Complex for the user, easy for them to make a mistake
2. Use a configuration technique that persists across upgrades
   * Pro: Nothing special for us to do
   * Con: Chick and egg problem for installation (incl race conditions)
   * Con: More complex for us to implement
   * Con: Discarding the approach we have already implemented
3. Store our configuration in a config map and merge it after upgrade
   * Similar to the way Helm stashes a copy of installed CRDs for later use
   * Con: Might be surprising to users
   * Con: Fragile - things go awry if user deletes the config map item
5. Query for existing CRDs and always support those
   * Pro: Seemless upgrades (Requirements #2, #3)
   * Pro: consistent with no-CRD Deletion (Requirement #4)
   * Pro: Installed resources never stop working
   * Pro: We can use existing commandline configuration
   * Pro: hard for users to get wrong
   * Pro: automatically works when users manually deploy CRDs

super-harsh commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,174 @@
+data:
+  "all": ""  # magic key meaning "*" which isn't supported in keys
+  microsoft.resources: "*"
+  microsoft.storage: "StorageAccount,StorageAccountsBlobService"
```
We might need handling somewhere to check if parent CRD is installed or not. As for in this example, what if a new user, does something like `microsoft.storage: "StorageAccountsBlobService"` and assume it'll include StorageAccount or if its not required.

matthchr replied:
I'll add an open questions section and put this there. My current thinking on this is that we avoid this complexity and strongly recommend users install entire providers such as `compute/*`.

There are other links to consider too:
- Owners (i.e. resourceGroup owns a lot of things even though it's not in their group)
- Cross resource references: i.e. compute references networking. This is pretty involved to work out automatically and there's also no guarantee that just because a resource _can_ link to a Subnet or a LoadBalancer that the customer actually has that use-case.

matthchr commented:
> A good write up of the problem, but I disagree with your conclusion.
> 
> I think your answers for Question 2 are actually trying to answer two different questions:
> 
> * How do users configure which CRDs to install
> * How do we ensure consistent configuration as ASO is upgraded?
> 
> For the later question, there are at least four answers:
> 
> 1. Require the user to be consistent when installing any new version
>    
>    * Pro: Easy for us
>    * Con: Complex for the user, easy for them to make a mistake
> 2. Use a configuration technique that persists across upgrades
>    
>    * Pro: Nothing special for us to do
>    * Con: Chick and egg problem for installation (incl race conditions)
>    * Con: More complex for us to implement
>    * Con: Discarding the approach we have already implemented
> 3. Store our configuration in a config map and merge it after upgrade
>    
>    * Similar to the way Helm stashes a copy of installed CRDs for later use
>    * Con: Might be surprising to users
>    * Con: Fragile - things go awry if user deletes the config map item
> 4. Query for existing CRDs and always support those
>    
>    * Pro: Seemless upgrades (Requirements [initial pr template draft #2](https://github.com/Azure/azure-service-operator/pull/2), #3)
>    * Pro: consistent with no-CRD Deletion (Requirement [EventHub Service Operator #4](https://github.com/Azure/azure-service-operator/issues/4))
>    * Pro: Installed resources never stop working
>    * Pro: We can use existing commandline configuration
>    * Pro: hard for users to get wrong
>    * Pro: automatically works when users manually deploy CRDs

I'm not sure that I follow what the difference is between 2 and 3. Isn't 2 just a generic version of 3?

On 4: Agree this seems like a great option. I've updated the document to include it, although I did remove this pro:
```
   * Pro: consistent with no-CRD Deletion
```

As every option has this pro (we never issue CRD delete no matter what so it feels weird to call it out explicitly here when we haven't anywhere else).

theunrepentantgeek commented:
> I'm not sure that I follow what the difference is between 2 and 3. Isn't 2 just a generic version of 3?

Option 2 is suggesting that we specifically use a form of configuration that persists across upgrades.

We could use a config map entry or a custom resource, but if the user overwrites it (either deliberately or accidentally) we would have the failure mode of orphaning existing resources.

Option 3 is suggesting that we make a copy of the configuration (regardless of the form) and stash it away somewhere so that we can reference it later, comparing it with what we're given later.

We can then compare/contrast/merge user provided configuration with our "memory" of what happened last time to ensure things run smoothly - or that we give better diagnostics.

matthchr commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,248 @@
+* More complex for us to implement, especially with Helm. Helm owns any resource it creates, 
+  and deletes it if not specified on the next upgrade. See [Using a ConfigMap](#answer-2b--use-a-configmap) for some 
+  workarounds.
+
```
Add variant: "Store the configuration independently..."

matthchr commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,248 @@
+
+### Answer to [question #3: How do we ensure consistent configuration as ASO is upgraded?](#question-3--how-do-we-ensure-consistent-configuration-as-aso-is-upgraded)
+
+[Answer 3c: Query existing CRDs and always support those](#answer-3c--query-existing-crds-and-always-support-those)
```
Note: We need to deal with the users-manage-their-own CRDs case, explain what we do for that here.

matthchr replied:
Confirm that in the "we dont manage CRDs case, we still have GET CRDs permissions"

matthchr replied:
Updated documentation to cover this case. Fundamentally, it should work exactly as planned. See the document which outlines the basic algorithm

matthchr commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,248 @@
+* Installed resources are always upgrade to the latest version, even if user has forgotten about them.
+* Could be built onto existing cmdline configuration.
+* Hard for users to get wrong.
+
```
+ Single source of truth

matthchr commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,248 @@
+
+## Status
+
+Proposed.
```
We need to document:
1. What the CRD uninstall process is (especially for users from v2.0.0 where they have everything).
2. Explain how to actually select new CRDs
3. Explain how CRDs are managed/installed.

matthchr replied:
Have made notes of all this. Documentation will be included in the implementation PR.

matthchr commented on docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md:
```diff
@@ -0,0 +1,248 @@
+
+## Status
+
+Proposed.
```
Open question, even without a predefined set of "default" resources, do we want a "default" set in the ASO install command on the readme so that people know what to do?

Consensus seems to be: yes.

matthchr replied:
Made a note of this. Docs will be in the implementation PR

matthchr commented:
/ok-to-test sha=d3e3666
